File: ai-services/app/models/impact.py
"""API contracts for the impact predictor — SOURCE OF TRUTH (mirror in web/lib/types).

The request carries the 6 model features the table was trained on (Phase 1.3). Mapping
barangay rows -> these features is Phase 4 (pipeline wiring), not this service.
"""
from typing import Literal, Optional

from pydantic import BaseModel, ConfigDict, Field


class BarangayFeatures(BaseModel):
    """One feature row, matching the Phase 1.3 training table's model inputs."""

    category_ordinal: int = Field(
        ..., ge=0, le=5, description="PAGASA intensity: 0 TD .. 5 violent typhoon"
    )
    total_houses: int = Field(..., ge=0, description="houses in the affected area")
    province_housing_units: int = Field(..., ge=0, description="HDX scale proxy")
    province_households: int = Field(..., ge=0, description="HDX scale proxy")
    structural_vuln_frac: float = Field(
        ..., ge=0.0, le=1.0, description="1 - strong-roof&wall share of housing"
    )
    unimproved_water_frac: float = Field(
        ..., ge=0.0, le=1.0, description="share on natural/peddler water sources"
    )
    id: Optional[str] = Field(None, description="caller key, echoed back; unused by the model")


class PredictImpactRequest(BaseModel):
    features: list[BarangayFeatures] = Field(..., min_length=1)


class ImpactPrediction(BaseModel):
    """One prediction, aligned by order to the request's `features`."""

    affected: int = Field(..., ge=0, description="predicted affected population")
    affected_confidence: float = Field(..., ge=0.0, le=1.0)

    # Reported in regressor framing; None in classifier framing.
    damage_rate: Optional[float] = Field(None, ge=0.0, le=1.0)
    damage_rate_confidence: Optional[float] = Field(None, ge=0.0, le=1.0)

    # Reported in classifier framing; None in regressor framing.
    severity_class: Optional[Literal["low", "moderate", "high", "severe"]] = None
    severity_confidence: Optional[float] = Field(None, ge=0.0, le=1.0)

    confidence: float = Field(..., ge=0.0, le=1.0, description="overall = min of reported heads")
    source: Literal["tabpfn", "heuristic"]
    id: Optional[str] = None


class PredictImpactResponse(BaseModel):
    model_config = ConfigDict(protected_namespaces=())

    predictions: list[ImpactPrediction]
    model_framing: Literal["regressor", "classifier"]
    latency_ms: float

File: web/lib/types/impact.ts
// Impact predictor API contract.
// MIRRORS ai-services/app/models/impact.py (the canonical source). Any change to one
// must update the other in the SAME commit (see CLAUDE.md > Rules).

export type ModelFraming = "regressor" | "classifier";
export type SeverityClass = "low" | "moderate" | "high" | "severe";
export type PredictionSource = "tabpfn" | "heuristic";

/** One feature row, matching the Phase 1.3 training table's model inputs. */
export interface BarangayFeatures {
  /** PAGASA intensity: 0 TD .. 5 violent typhoon. */
  category_ordinal: number;
  total_houses: number;
  province_housing_units: number;
  province_households: number;
  /** 1 − strong-roof&wall share of housing, [0,1]. */
  structural_vuln_frac: number;
  /** share on natural/peddler water sources, [0,1]. */
  unimproved_water_frac: number;
  /** caller key, echoed back; unused by the model. */
  id?: string | null;
}

export interface PredictImpactRequest {
  features: BarangayFeatures[];
}

/** One prediction, aligned by order to the request's `features`. */
export interface ImpactPrediction {
  affected: number;
  affected_confidence: number;
  /** 80% predictive interval for `affected` (0.1/0.9 quantiles). Null on the heuristic
   *  path (no interval). Decision support, not a guarantee. */
  affected_low: number | null;
  affected_high: number | null;
  /** Set in regressor framing; null in classifier framing. */
  damage_rate: number | null;
  damage_rate_confidence: number | null;
  /** 80% predictive interval for `damage_rate`; regressor framing only. */
  damage_rate_low: number | null;
  damage_rate_high: number | null;
  /** Set in classifier framing; null in regressor framing. */
  severity_class: SeverityClass | null;
  severity_confidence: number | null;
  /** Overall = min of the reported heads. */
  confidence: number;
  source: PredictionSource;
  id?: string | null;
}

export interface PredictImpactResponse {
  predictions: ImpactPrediction[];
  model_framing: ModelFraming;
  latency_ms: number;
}
